rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+- `lessThan`: Compares left < right.
+- `greaterThanOrEqual`: Compares left >= right
+- `lessThanOrEqual`: Compares left <= right
+- `regexMatch`: Whether a given value matches the given regular expression
+
+#### Value Expressions
+
+The left and right side of a comparison expression are value expressions which can either be:
+- Literals
+- Session Variable references
+
+In the future, these may be expanded to:
+- The output from an invoked command
+- The result of invoking an external policy engine like Open Policy Agent (OPA).
+
+#### Example
+
+Here is a rule that descriebs the condition to enable an authorization rule if the user is an admin or is an unbanned user in the PRO tier.
+
+```yaml
+or:
+  - and:
+    - equal:
+        left:
+          sessionVariable: customer_tier
+        right:
+          literal: PRO_TIER
+    - not:
+        equal:
+          left:
+            sessionVariable: is_banned
+          right:
+            literal: true
+  - equal:
+      left:
+        sessionVariable: is_admin
+      right:
+        literal: true
+```
+
+### Type Permissions
+
+The following permission primitives are available when defining a permissions rule for a type.
+- Allow access to certain fields
+- Deny access to certain fields
+
+If a field is both allowed and denied, it will be denied.
+
+#### Example
+
+```yaml
+kind: TypePermissions
+version: v2
+definition:
+  typeName: Order
+  authorizationRules:
+    # Rule 1: Allow all fields by default
+    - allowFields: "*"
+      condition:
+        literal: true
+    # Rule 2: Deny internal_order_notes to non-admins. Deny always takes precedence over allow.
+    - denyFields:
+        - internal_order_notes
+      condition:
+        not:
+          equals:
+            left: sessionVariable: x-hasura-role
+            right: literal: admin
+```
+
+### Model Permissions
+
+The following permission primitives are available when defining a permissions rule for a model.
+- Allow access to certain objects or "rows" (defined by a predicate)
+- Deny access to certain objects or "rows" (defined by a predicate)
+
+If an object/row is both allowed and denied, it will be denied.
+
+#### Example
+
+```yaml
+kind: ModelPermissions
+version: v2
+definition:
+  modelName: Orders
+  authorizationRules:
+    # Rule 1: Allow users to see their own orders
+    - allowObjects:
+        fieldComparison:
+          fieldName: user_id
+          operator: _eq
+          value:
+            sessionVariable: x-hasura-user-id
+      condition:
+          equals:
+            left: sessionVariable: x-hasura-role
+            right: literal: user
+    # Rule 2: Disallow anyone except admins to see hidden orders
+    - denyObjects:
+        fieldComparison:
+          fieldName: is_hidden
+          operator: _eq
+          value: 
+            literal: true
+      condition:
+        not:
+          equals:
+            left: sessionVariable: x-hasura-role
+            right: literal: admin
```

Comment:
I think the idea is that (for now at least) expressions are evaluated in the engine only, since they only depend on the values of session variables. I agree that pushing down subexpressions is not a tractable plan.